packages/gateway/src/gateway.js

```diff
@@ -0,0 +1,133 @@
+/* eslint-env serviceworker, browser */
+
+import pAny from 'p-any'
+import pSettle from 'p-settle'
+
+import { getCidFromSubdomainUrl } from './utils/cid.js'
+
+/**
+ * @typedef {Object} GatewayResponse
+ * @property {Response} response
+ * @property {string} url
+ * @property {number} responseTime
+ */
+
+/**
+ * Handle gateway request
+ *
+ * @param {Request} request
+ * @param {import('./env').Env} env
+ * @param {import('./index').Ctx} ctx
+ */
+export async function gatewayGet(request, env, ctx) {
+  const reqUrl = new URL(request.url)
+  const cid = getCidFromSubdomainUrl(reqUrl)
+
+  const gatewayReqs = env.ipfsGateways.map(async (url, index) => {
+    const ipfsUrl = new URL('ipfs', url)
+    const controller = new AbortController()
+    const startTs = Date.now()
+    const timer = setTimeout(() => controller.abort(), env.REQUEST_TIMEOUT)
+
+    const response = await fetch(
+      `${ipfsUrl.toString()}/${cid}${reqUrl.pathname || ''}`,
+      { signal: controller.signal }
+    )
+
+    clearTimeout(timer)
+    /** @type {GatewayResponse} */
+    const gwResponse = {
+      response,
+      url,
+      responseTime: Date.now() - startTs,
+    }
+    return gwResponse
+  })
+
+  try {
+    const { response, url } = await pAny(gatewayReqs)
+
+    ctx.waitUntil(
+      (async () => {
+        const responses = await pSettle(gatewayReqs)
+        const successFullResponses = responses.filter(
+          (r) => r.value?.response?.ok
+        )
+
+        await Promise.all([
+          updateMetrics(request, env, responses, url),
+          updateCidsTracker(request, env, successFullResponses, cid),
+        ])
+      })()
+    )
+
+    // forward gateway response
+    return response
+  } catch (err) {
+    // Update metrics as all requests failed
+    const responses = await pSettle(gatewayReqs)
+    await updateMetrics(request, env, responses)
+    throw err
+  }
+}
+
+/**
+ * @param {Request} request
+ * @param {import('./env').Env} env
+ * @param {pSettle.PromiseResult<GatewayResponse>[]} responses
+ * @param {string} [fasterUrl]
+ */
+async function updateMetrics(request, env, responses, fasterUrl) {
+  const id = env.metricsDurable.idFromName('metrics')
```

Comment:
Can this name be a constant? Also, is this right? Isn't it "METRICS" and "CIDSTRACKER"?